### PR TITLE
Fix error handling using auth redirect flow

### PR DESCRIPTION
**Pull request by stephenglass**
### Background

Fixes issue #24856 which causes the user to be stuck on a blank screen if there is any error in authentication process or exception thrown in a custom a `signInResolver` or `profileTransform` when using auth redirect flow enabled by `enableExperimentalRedirectFlow` config. There is only code to handle errors when using popup flow but none when using the redirect flow hence why nothing happens and user is stuck on the blank screen.

I have a requirement to use auth redirect flow and have an email based allowlist for accessing Backstage which is implemented using a custom `profileTransform` in the auth provider. Having a user stuck on the blank page without letting them know the error reason won't be accessible.

### Solution

1. If there is an error caught in authentication and using redirect flow, redirect the user back to app origin with `error` URL parameter contain the error message.
2. Sign-in pages can use the error message from the URL param to display on the page

Pics on displaying the error (same as it would show using popup flow):

Single Provider
image

Multi Providers
image

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by benjdlambert** [comment]
Hey :wave: sorry this has taken a while to get around to :pray:

What's the motivation for cookie over querystring? I appreciate its pretty ugly, but it's a pretty common pattern with authenticating proxies, and trying to understand the pros/cons to make a better decision on the best path forward.

**Comment by stephenglass** [comment]
> Hey 👋 sorry this has taken a while to get around to 🙏
> 
> What's the motivation for cookie over querystring? I appreciate its pretty ugly, but it's a pretty common pattern with authenticating proxies, and trying to understand the pros/cons to make a better decision on the best path forward.

Thanks for taking a look @benjdlambert !

1. **User experience**: In addition to it just not looking nice and possibly confusing, what if a user shares `http://mybackstage.net/?error="{message:'hello'}''` link to a co-worker. Then the error message would also show to the other user when there is actually no error. What if user refreshes the page with the error message - it would keep showing when most users would expect it to be cleared after a refresh. With enough error redirects, browsers may also cache and suggest the url with error query params when the user types into the address bar (this is an annoyance with sharepoint and servicenow). I could probably go on about the UX.
2. **Security**: Storing error message in a `httpOnly` cookie is not accessible with JS which reduces risk of XSS attacks. Error message in url can be manipulated by user and could lead to vulnerability if not properly sanitized. The error messages in query param url will also be logged in browser history which is bad if there is potentially sensitive data in the error message. There are also url length limits to be mindful of if users have large error messages. 

Imo, using a `httpOnly` cookie adds complexity compared to a stateless solution like url query params but offer better UX, security, and potentially some easier scalability if the auth flow changes over time.

**Comment by stephenglass** [comment]
Hey @benjdlambert any update on if this will be an acceptable approach?

**Comment by benjdlambert** [comment]
@stephenglass sorry - haven't forgotten, let me bring this up internally today and get back to you!

**Review by Rugvip — COMMENTED** [review]
Yep I see the benefits on the cookie flow, but it does introduce a lot of complexity 😅. It's a bit extra unfortunate due to the redirect flow being a non-standard flow for sign-in.

The first thing that comes to mind for me is if we could solve most of the same issues with ´location.replace(...)´? That might avoid the history entry and the nuisance of auto-completing to weird URLs. So we'd return the full error message, but the frontend would immediate grab the error message, render it, and then remove it from the location using `location.replace(...'` or `react-router` equivalent.

When it comes to the sign-in error API, I'm not sure there's too much benefit to having that be a utility API tbh? Perhaps that could just be a function exported for re-use if we feel that it's needed in the end

**Comment by stephenglass** [comment]
> Yep I see the benefits on the cookie flow, but it does introduce a lot of complexity 😅. It's a bit extra unfortunate due to the redirect flow being a non-standard flow for sign-in.
> 
> The first thing that comes to mind for me is if we could solve most of the same issues with ´location.replace(...)´? That might avoid the history entry and the nuisance of auto-completing to weird URLs. So we'd return the full error message, but the frontend would immediate grab the error message, render it, and then remove it from the location using `location.replace(...'` or `react-router` equivalent.
> 
> When it comes to the sign-in error API, I'm not sure there's too much benefit to having that be a utility API tbh? Perhaps that could just be a function exported for re-use if we feel that it's needed in the end


@Rugvip  I've tried messing with this today and was not successful in manipulating the history to be single entry without the error message. As for the sign-in error API, I agree with you and that's a great idea. I've pushed some changes that removes the API and instead creates a new `useSignInAuthError` hook in the `auth-react` plugin. I think it makes the contribution a little more straightforward.

**Comment by stephenglass** [comment]
@Rugvip @benjdlambert any chance this can get another review? 🙏

**Comment by benjdlambert** [comment]
@stephenglass sorry it's taking a while to get this through - @Rugvip is out until next week, will pick it up again when he's back next week if that's OK? :pray:

**Comment by stephenglass** [comment]
@benjdlambert great, thank you for the update. Just wanting to make sure this is in the queue 👍

**Comment by stephenglass** [comment]
Hey @Rugvip just checking in to see when this will be able to get picked up to review?

**Comment by Rugvip** [comment]
> > Yep I see the benefits on the cookie flow, but it does introduce a lot of complexity 😅. It's a bit extra unfortunate due to the redirect flow being a non-standard flow for sign-in.
> > The first thing that comes to mind for me is if we could solve most of the same issues with ´location.replace(...)´? That might avoid the history entry and the nuisance of auto-completing to weird URLs. So we'd return the full error message, but the frontend would immediate grab the error message, render it, and then remove it from the location using `location.replace(...'` or `react-router` equivalent.
> > When it comes to the sign-in error API, I'm not sure there's too much benefit to having that be a utility API tbh? Perhaps that could just be a function exported for re-use if we feel that it's needed in the end
> 
> @Rugvip I've tried messing with this today and was not successful in manipulating the history to be single entry without the error message. As for the sign-in error API, I agree with you and that's a great idea. I've pushed some changes that removes the API and instead creates a new `useSignInAuthError` hook in the `auth-react` plugin. I think it makes the contribution a little more straightforward.

Alright, I think we'll just stick to URL parameters that show up to the user then. We want to avoid adding too much complexity to support this.

**Comment by stephenglass** [comment]
> > > Yep I see the benefits on the cookie flow, but it does introduce a lot of complexity 😅. It's a bit extra unfortunate due to the redirect flow being a non-standard flow for sign-in.
> > > The first thing that comes to mind for me is if we could solve most of the same issues with ´location.replace(...)´? That might avoid the history entry and the nuisance of auto-completing to weird URLs. So we'd return the full error message, but the frontend would immediate grab the error message, render it, and then remove it from the location using `location.replace(...'` or `react-router` equivalent.
> > > When it comes to the sign-in error API, I'm not sure there's too much benefit to having that be a utility API tbh? Perhaps that could just be a function exported for re-use if we feel that it's needed in the end
> > 
> > 
> > @Rugvip I've tried messing with this today and was not successful in manipulating the history to be single entry without the error message. As for the sign-in error API, I agree with you and that's a great idea. I've pushed some changes that removes the API and instead creates a new `useSignInAuthError` hook in the `auth-react` plugin. I think it makes the contribution a little more straightforward.
> 
> Alright, I think we'll just stick to URL parameters that show up to the user then. We want to avoid adding too much complexity to support this.

@Rugvip @benjdlambert Updated the code to use URL parameters instead of cookies - its a lot more simple now. Also updated the PR description with new screenshots. Appreciate getting another check when you get a chance, thanks!

File: packages/core-components/src/layout/SignInPage/providers.tsx [hunk 1]
```diff
@@ -88,6 +93,22 @@ export const useSignInProviders = (
...
+  const { t } = useTranslationRef(coreComponentsTranslationRef);
+  // User was redirected back to sign in page with error from auth redirect flow
+  const [searchParams, _setSearchParams] = useSearchParams();
+
+  useMountEffect(() => {
+    const errorParam = searchParams.get('error');
+    if (errorParam) {
+      errorApi.post(
+        new ForwardedError(t('signIn.loginFailed'), {
+          name: 'Error',
```
Comment (Rugvip) [inline packages/core-components/src/layout/SignInPage/providers.tsx:105]: Could you make the second arg a regular error object? `new Error(decodeURIComponent(errorParam))`

File: plugins/auth-node/src/oauth/createOAuthRouteHandlers.ts [hunk 2]
```diff
@@ -248,11 +249,20 @@ export function createOAuthRouteHandlers<TProfile>(
...
           : new Error('Encountered invalid error'); // Being a bit safe and not forwarding the bad value
-        // post error message back to popup if failure
-        sendWebMessageResponse(res, origin, {
-          type: 'authorization_response',
-          error: { name, message },
-        });
+
+        if (state?.flow === 'redirect' && state?.redirectUrl) {
+          const redirectUrl = new URL(state.redirectUrl);
+          redirectUrl.searchParams.set('error', encodeURIComponent(message));
```
Comment (Rugvip) [inline plugins/auth-node/src/oauth/createOAuthRouteHandlers.ts:255]: `searchParams` already encodes the value, so I think you're double encoding here and double decoding in the frontend

**Review by Rugvip — CHANGES_REQUESTED** [review]
Nice, thank you! 👍 

Just some small bits

**Comment by stephenglass** [comment]
@Rugvip Any chance this can make it to the upcoming main or next release? 🤞

**Review by Rugvip — APPROVED** [review]
Thank you, let's :shipit:! 🎉

**Pull request merged**